src/Storages/RocksDB/EmbeddedRocksDBBulkSink.cpp

```diff
@@ -0,0 +1,243 @@
+{
+    try
+    {
+        if (fs::exists(insert_directory_queue))
+            fs::remove_all(insert_directory_queue);
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__, fmt::format("Error while removing temporary directory {}:", insert_directory_queue));
+    }
+}
+
+std::vector<Chunk> EmbeddedRocksDBBulkSink::squash(Chunk chunk)
+{
+    /// End of input stream
+    if (chunk.getNumRows() == 0)
+    {
+        if (chunks.empty())
+            return {};
+        std::vector<Chunk> to_return;
+        std::swap(to_return, chunks);
+        return to_return;
+    }
+
+    /// Just read block is already enough.
+    if (isEnoughSize(chunk))
+    {
+        /// If no accumulated data, return just read block.
+        if (chunks.empty())
+        {
+            chunks.emplace_back(std::move(chunk));
+            return {};
+        }
+
+        /// Return accumulated data (maybe it has small size) and place new block to accumulated data.
+        std::vector<Chunk> to_return;
+        std::swap(to_return, chunks);
+        chunks.emplace_back(std::move(chunk));
+        return to_return;
+    }
+
+    /// Accumulated block is already enough.
+    if (isEnoughSize(chunks))
+    {
+        /// Return accumulated data and place new block to accumulated data.
+        std::vector<Chunk> to_return;
+        std::swap(to_return, chunks);
+        chunks.emplace_back(std::move(chunk));
+        return to_return;
+    }
+
+    chunks.emplace_back(std::move(chunk));
+    if (isEnoughSize(chunks))
+    {
+        std::vector<Chunk> to_return;
+        std::swap(to_return, chunks);
+        return to_return;
+    }
+
+    /// Squashed block is not ready.
+    return {};
+}
+
+std::pair<ColumnString::Ptr, ColumnString::Ptr> EmbeddedRocksDBBulkSink::serializeChunks(const std::vector<Chunk> & input_chunks) const
+{
+    auto serialized_key_column = ColumnString::create();
+    auto serialized_value_column = ColumnString::create();
+
+    {
+        auto & serialized_key_data = serialized_key_column->getChars();
+        auto & serialized_key_offsets = serialized_key_column->getOffsets();
+        auto & serialized_value_data = serialized_value_column->getChars();
+        auto & serialized_value_offsets = serialized_value_column->getOffsets();
+        WriteBufferFromVector<ColumnString::Chars> writer_key(serialized_key_data);
+        WriteBufferFromVector<ColumnString::Chars> writer_value(serialized_value_data);
+
+        for (const auto & chunk : input_chunks)
+        {
+            const auto & columns = chunk.getColumns();
+            auto rows = chunk.getNumRows();
+            for (size_t i = 0; i < rows; ++i)
+            {
+                for (size_t idx = 0; idx < columns.size(); ++idx)
+                    serializations[idx]->serializeBinary(*columns[idx], i, idx == primary_key_pos ? writer_key : writer_value, {});
+                writeChar('\0', writer_key);
+                writeChar('\0', writer_value);
```

Comment:
@vdimir the writer buffer in two places are being used differently:

- In `EmbeddedRocksDBSink`, it is used to serialised one key-value pair then to call `WriteBatch`.
- In `EmbeddedRocksDBBulkSink`, it is used to serialised keys and values to two `ColumnString`. That's why after serialising I added trailing zeros, because row in `ColumnString` should be null-terminated.

I wil check more but it's quite certain that the two serialisation logic are same, because in `ColumnString` the last null character is not part of the string. I will add a test for it as well (e.g. first import some key-value pairs not using bulk-sink, then enable bulk-sink and import pairs with the key but different value, and make sure that the later values override the prior).

> Actually, it might be good to unify the serialization code between EmbeddedRocksDBSink and EmbeddedRocksDBBulkSink so that it's clear they are serialized in the same way.

I think we can change the `EmbeddedRocksDBSink` logic, not serialise -> write batch -> serialise -> ... but serialise all records first, then write batch. RocksDB write batch has API to write multiple key-value pairs at once. Wdyt?

Reply:
> I think we can change the EmbeddedRocksDBSink logic, not serialise -> write batch -> serialise -> ... but serialise all records first, then write batch. RocksDB write batch has API to write multiple key-value pairs at once. Wdyt?

Wouldn't it increase memory consuption since we will have to have whole batch serialized to string in memory instead of single row in `WriteBufferWithOwnString`

Reply:
Yes, that's also my concern. But it's the only approach if we want to unify the serialisation logic in those two class.

May be as you said, we can left it for future improvement. I will just add a test for now.

Reply:
Added test to compare bulk sink vs normal sink serialisation.